thinc/backends/_custom_kernels.py

```diff
@@ -5,7 +5,12 @@
 
 try:
     import cupy
-except ImportError:
+
+    try:
+        cupy.cuda.runtime.getDeviceCount()
+    except cupy.cuda.runtime.CUDARuntimeError:
+        cupy = None
```

Comment:
I am not sure I understand this check. If there are no actual devices, wouldn't/shouldn't we fail in `CupyOps` already?

Reply:
The `CupyOps` module imports dependencies from `_custom_kernels`, and none of the `cupy`-related imports in the former raise an exception when there are no actual devices. The first-chance exception gets raised only when an actual `cupy` function gets called, i.e., when the `KERNELS.get_function(name)` function from the latter module gets called.

To be fair, perhaps the above method isn't the best way to catch these errors after all. We have a bunch of modules that perform their own flavour of compatibility checks, leading to multiple sources of truth. I think it would be better to do all the compatibility-related checks in a single module (`compat.py` as Sofie suggested) and then use that as the source of truth in all other modules. If we wish to retain backward-compatibility, we can also re-export the necessary variables (`has_cupy`, etc) from those modules where we're currently doing this. Although in my opinion, we probably want to avoid re-exports as it unnecessarily diffuses/increases the public API's contract/surface area.

Does that sound like a reasonable approach?

Reply:
Sounds good!